### DISCUSSION TITLE
Web URL

### DISCUSSION THREAD

**Discussion by beedawg1985** [h-a-graham/EAlidaR#22] [General]
Hey, nice package! It's my understanding that the tile token URLs are generated dynamically by the request 'Get Available Tiles' button and relate to the area previously defined (by polygon or drawing etc.) They will expire after some time, I think. 

I've written a similar script here: https://github.com/barnabasharris/rspatial/blob/master/getLiDAR.R it would be good to combine forces somehow

**Comment by h-a-graham** [comment]
Hey, just had a quick look through your script, looks great!, Wish I'd seen this sooner Awesome that you got this working with r selenium, I struggled with this (hence why I used python for this bit) so will be interesting to see what I missed. Yes, as you can probably see from the issues, my idea of saving all the arc map IDs for each tile is not a good solution,  obviously the best solution would be for the ea to provide stable urls for each tile but I've now been waiting over 2 weeks for a question on the defra data forum to be answered and now it seems they're avoiding answering the questions on twitter. 

In principal, what do you think about using your r selenium approach as a work around for now so that each time a request is made new arc IDs are generated? Is this a stable solution, do you think? 
Part of the reason I avoided an approach like this to begin with was speed of download for many tiles (though this will still be much quicker than the portal and slower is better than not working!), I'll do some testing with your approach next week. 

One of the key issues I had with r selenium was the need for docker? Did you get around this? The need for docker might make the package a little less user friendly. For example, I had real trouble installing docker on my work pc. If docker is necessary for r selenium is it safer to use python which is perhaps easier for people to set up? 

Really appreciate you getting in touch on this, I'd sort of given up on finding a good work around but I really think your approach provides a good solution. Thanks!

**Reply by h-a-graham** [reply]
Couldn't help but look a bit more at your script. Huge thing is that I'd been sending keys to the wrong element! So thanks for that. I think there is actually a very straight forward solution using your approach. Question I think now is which of the following approach is preferable as a package.
1. Use reticulate to run python selenium.  Set up requires python install, install selenium and download gecko driver then things should easily run out of the box after an initial overhead at the set up/installation.

2. Install docker which needs to be opened each time the package is used.

If there are other options please let me know. I think I'll have a go at implementing option 1. Right now, this involves the least work so might be worth a spin. Let me know your thoughts.

Thanks again

**Comment by beedawg1985** [comment]
Heyup, 
Good to hear from you, I'll try and address your points in order:

1) R Selenium; I think it would make sense to use this package over python as it helps keeps the package 'pure-R-ish' / reduce dependencies, though I don't think it can ever run truly 'out of the box', due to the neccessary installation of external Selenium libraries. Until an ESRI / EA provide an API this will always be the case I suppose. I don't think the selenium server is dependent on Docker when called using the R Selenium `rsDriver` function, see (https://cran.r-project.org/web/packages/RSelenium/vignettes/basics.html#do-i-need-to-run-a-selenium-server), though I'm not really sure what is happening under the hood here to be honest. I'll have a dig about and try and remember how I got it running on my Ubuntu box.

2) Download URLs; my own experiments seemed to indicate that a request through the portal generated token-based link(s) to the requested tile(s), which eventually expire. I have previously come across repos on GitHub which listed scraped URLs for various tiles but none of them worked when I tried them. The only viable solution I found was to generate new requests on the fly. I'm still yet to work out the upper limit per request (seems to be a certain number of square kilometres but also limited by number of grid squares possibly?).

P.S.
If you try using my script make sure you switch the comments on the RsDriver call, so:
```` # define chrome options
  eCaps <- list(chromeOptions = list(
    args = c(
      '--disable-gpu'
      ,'--headless',
      '--window-size=1280,800'
    )
  ))
  rD <- rsDriver(browser = "chrome",
                 chromever = "81.0.4044.138",
                 extraCapabilities = eCaps,
                 port =
                   as.integer(base::sample(seq(32768,65535, by=1),1)))
  # rD <- RSelenium::rsDriver(
  #   browser = "firefox",
  #   extraCapabilities = list(
  #     "moz:firefoxOptions" = list(
  #       args = list('--headless')
  #     )
  #   ),
  #   port = 
  #     as.integer(base::sample(seq(32768,65535, by=1),1))
  # )

**Comment by h-a-graham** [comment]
Okay this is great to hear I'll see if I can get something working with R selenium - totally agree this is the way to go if we can avoid Docker (have to admit, selenium is a bit of a new venture for me and I rushed this step initially assuming that the tokens might last longer than they did). They python selinium scraping approach that I worked up works pretty well - basically uploads 10km grid square and scrapes the id/token. I did this for all 10km grids in england but I think it would be easy t repurpose so, as you say, they are generated at the point of download. So my thinking is spatial join wiht input polygon and 10km grid, retirieve tokens, then 10km join with 5km grid and generate URLs. The tokens I beleive are arc web map IDs which are have spatial meaning so a code generated for a 10km grid will be valid for urls on all four 5km grids that are contained. I think this is the simplest approach, though I'm sure we could wventually come up with a more efficent approach to minimise the uploads to the portal. I'll give this a whirl and then  let you know how it goes. I'm fairly optimistic about this approach. Thanks again for your help - really appreciated!

**Reply by beedawg1985** [reply]
Hey yep that sounds all good. I didn't realise the tokens worked in that way, that's pretty useful to know.

I think I understand your  approach using the 5km/10km squares — how would it work when someone wanted just a small (say sub 5km), irregular polygon of LiDAR? Would the current approach request links for all 5km tiles falling within whichever 10km grid square the polygon fell within? 

For what its worth I reckon it would be super useful to have multiple request modes;

1) text based grid square request ('SENE', 'SYSE', 'SU') (with large request areas being split up into smaller request that don't trip the max limit).
2) Point mode (eastings + northings are used to generate a user defined buffer polygon which forms the request)
3) Polygon mode, using a simple features polygon, which allows users to define the exact area they require. This is the how the script I wrote works, but it does currently rely on some *nix specific system calls  which would need to be rewritten to be cross-platform compatible. 

P.S Just comes across this thread: 
https://stackoverflow.com/questions/33741921/can-i-run-multiple-instances-at-oncesimultaneously-with-selenium-webdriver
Might be cool for speeding up big requests that have been split down into smaller requests!

**Comment by h-a-graham** [comment]
Hey, 

So, in the example you mention, a single 10km tile is uploaded and the token/id is scraped. This will then work for any of the four 5km tiles within that 10km grid thus reducing the number of uploads. so shouldn't be a time penalty in this case.  

re your suggestions:
1. I agree with the idea of multiple download options. My previous appraoch allowed a single tile download which was handy but might need rethinking for this new appraoch. 
2. This is a great idea and straight forward to implement.
3. This I think should be the main function which maybe functions 1 and 2 above could call as wrappers. I suspect most people will use this option anyway so I'll prioritise this approach.

Yeah I like the idea of running it in parallel somehow - will be an interesting thing to try out - can't see why this couldn't work. I've had good experience with the foreach package and %dopar% and seems to be the most stable parallism option across platforms with R. I'll add that to the to do list once we get this working.

Also, I just stumbled across another potential thing to consider. Originally, I had only implemented downloads for 'LIDAR Composite DSM' and "LIDAR Composite DTM" products. Do you know if these differ much from "LIDAR Tiles DSM"     "LIDAR Tiles DTM" products? And further to this, the new data is now labelled 'National LIDAR programme DTM' and 'National LIDAR programme DSM' (Yay for consistency!). So how do you think we should handle this - I'd like to keep the  defaults as simple as possible so if a user requests something like:

`get_area(poly_area = Ashop_sf, resolution = 2, model_type = 'DTM', merge_tiles=TRUE, crop=TRUE)`  Then they are given the best (most recent) data. Given this, do you think it is okay to have default selection (for DTM): 
'National LIDAR programme DTM', "LIDAR Composite DTM", "LIDAR Tiles DTM" ?

FYI after some initial issues with the rsDriver (seemed to relate to Java path in windows) it is now working great with firefox but not chrome. 

Cheers!

**Reply by beedawg1985** [reply]
Heyup, 
That's all good; I think the 10km square request method sounds robust.

That's good you're familiar with foreach, it's not something I've used at all. I've used the `future.lapply ` function a handful of times. I suppose we have to weigh up increasing package dependencies against speeding up large requests.

The different EA products... I *think* the composite layers are products which includes all surveys of a set resolution, regardless of when they were conducted (so a single tile could contain surveys from multiple years). The tile products are broken down by year.
It's a cumbersome system but useful in some ways; I originally wrote the script I shared with you to find coverage locations with a minimum number of years of worth of survey (so I could construct time series datasets). You'll see in my script a 'min survey' argument which specifies this. I think the composite layer should eventually contain the new National programme data but for some reason it has its own product at the moment....? 

It might be worth amending the `model_type` argument to an actual product code (e.g. composite or national lidar programme) but equally I can see the appeal in keeping it simple.

Glad you got rsDriver running! Yeah I had some mix and match issues with chrome/firefox initially too.

**Reply by h-a-graham** [reply]
Okay, that's all super helpful, thanks!

As a quick update -  I have this working but it needs lots of tidying up still. Need to add in better error handling etc. Using selenium in this way is going to be a lot less stable but to be honest I can't see a way around this. The key here I think is going to be reporting erros correctly and automating retries where something weird happens. 

It is pretty slow too... some basic bench marking (with half decent copper connection) allows me to download 6 5km tiles (from 4 10km scrapes) in about 5 minutes. The spatial join in the portal is mind bendingly slow! I definitely think some parallel scraping would help with this problem. For now I think I'll get everything working as best as I can with a single core - making sure error handling is correct etc. Once this is improved I'm very keen to try out a parallel approach. Something which could work well is offering a `multi_core = TRUE` option which could use foreach or (another package ) for this but add it as an optional dependency where single core runs fine without it... That being said I think foreach is fairly lightweight and stable so I'm not too concerned about having it as a dependency.

re. the data options, I hadn't realised there were so many data types! I can totally understand the interest in the time series data so it would be awsome to implement this. For now I've implemented an approach where the National Lidar Programme (NLP) data is searched first and failing that the composite data is searched for. I like the idea of keeping `model_type = c('DTM, DSM)` but I think it would be sweet to add an option like 'Lidar_strategy = c('newest', 'composite', 'timeseries') where time series would return a raster stack of all available time steps?

Now have rsDriver working with Chrome, which I'm happy with for now - I hadn't grasped the need to specify your own chrome version rather than a provided driver. I'll keep working on this today but, if you're up for it - I'd be very grateful of some testing once I have something a bit further along?

Cheers!

**Comment by h-a-graham** [comment]
@beedawg1985 Thanks again for all you help with this. I've implemented several of the things we've discussed here and  pulled the changes to the master branch. Let me know if you have any issues with this or suggestions. Currently the download priorities NLP and then Composite. I will try to add support for the time series data soon but need to work on some other stuff for a bit. Cheers!
